ext/spl/spl_observer.c

```diff
@@ -361,6 +443,69 @@ PHP_METHOD(SplObjectStorage, attach)
 	spl_object_storage_attach(intern, obj, inf);
 } /* }}} */
 
+static int spl_object_storage_has_dimension(zend_object *object, zval *offset, int check_empty)
+{
+	spl_SplObjectStorage *intern = spl_object_storage_from_obj(object);
+	if (UNEXPECTED(offset == NULL || Z_TYPE_P(offset) != IS_OBJECT || (intern->flags & SOS_OVERRIDDEN_READ_DIMENSION))) {
+		/* Can't optimize empty()/isset() check if getHash, offsetExists, or offsetGet is overridden */
+		return zend_std_has_dimension(object, offset, check_empty);
+	}
+	spl_SplObjectStorageElement *element = zend_hash_index_find_ptr(&intern->storage, Z_OBJ_HANDLE_P(offset));
+	if (!element) {
+		return 0;
+	}
+
+	if (check_empty) {
+		return i_zend_is_true(&element->inf);
+	}
+	/* NOTE: SplObjectStorage->offsetExists() is an alias of SplObjectStorage->contains(), so this returns true even if the value is null. */
+	return 1;
+}
+
+static zval *spl_object_storage_read_dimension(zend_object *object, zval *offset, int type, zval *rv)
+{
+	spl_SplObjectStorage *intern = spl_object_storage_from_obj(object);
+	if (UNEXPECTED(offset == NULL || Z_TYPE_P(offset) != IS_OBJECT || (intern->flags & SOS_OVERRIDDEN_READ_DIMENSION))) {
+		/* Can't optimize it if getHash, offsetExists, or offsetGet is overridden */
+		return zend_std_read_dimension(object, offset, type, rv);
+	}
+	spl_SplObjectStorageElement *element = zend_hash_index_find_ptr(&intern->storage, Z_OBJ_HANDLE_P(offset));
+
+	if (!element) {
+		if (type == BP_VAR_IS) {
+			return &EG(uninitialized_zval);
+		}
+		zend_throw_exception_ex(spl_ce_UnexpectedValueException, 0, "Object not found");
+		return NULL;
+	} else {
+		/* This deliberately returns a non-reference, even for BP_VAR_W and BP_VAR_RW, to behave the same way as SplObjectStorage did when using the default zend_std_read_dimension behavior.
+		 * i.e. This prevents taking a reference to an entry of SplObjectStorage because offsetGet would return a non-reference. */
+		ZEND_ASSERT(Z_TYPE(element->inf) != IS_REFERENCE);
+		ZVAL_COPY(rv, &element->inf);
```

Comment:
I'm not sure why element of the storage can't be a reference. May be this correct. May be it's safer to use ``ZVAL_COPY_DEREF()``

Reply:
I don't believe it can be a reference - it's because you can't create a reference with offsetSet, offsetGet, and offsetExists in the zend_std_read_dimension code.

In many places it was using ZVAL_COPY.

Attempting to do so resulted in `Notice: Indirect modification of overloaded element of SplObjectStorage has no effect in php shell code on line 1` before/after this change

----

Separately, I think there might be a bug in the `__unserialize` implementation when a value is a reference
The handlers I've added are based on zend_std_read_dimension and zend_std_write_dimension. https://github.com/php/php-src/pull/7695#discussion_r757887810

Reply:
```php
php > echo PHP_VERSION;
7.4.19-dev
php > $s = new SplObjectStorage();
php > $y = 1;
php > $x = [new stdClass(), &$y];
php > var_dump($s->__unserialize([$x, []]));
NULL
php > var_dump($s);
object(SplObjectStorage)#1 (1) {
  ["storage":"SplObjectStorage":private]=>
  array(1) {
    ["000000001f5767c4000000002919f006"]=>
    array(2) {
      ["obj"]=>
      object(stdClass)#2 (0) {
      }
      ["inf"]=>
      &int(1)
    }
  }
}
```